cfg/builder/builder_walk.cc

```diff
@@ -496,46 +496,51 @@ BasicBlock *CFGBuilder::walk(CFGContext cctx, ast::Expression *what, BasicBlock
-                auto rescueStartBlock = cctx.inWhat.freshBlock(cctx.loops, cctx.rubyBlockId);
-                unconditionalJump(current, rescueStartBlock, cctx.inWhat, a->loc);
-                cctx.rescueScope = rescueStartBlock;
+                auto bodyRubyBlockId = ++cctx.inWhat.maxRubyBlockId;
+                auto handlersRubyBlockId = ++cctx.inWhat.maxRubyBlockId;
+                auto ensureRubyBlockId = ++cctx.inWhat.maxRubyBlockId;
+                auto elseRubyBlockId = ++cctx.inWhat.maxRubyBlockId;
+                auto rescueHeaderBlock = cctx.inWhat.freshBlock(cctx.loops, cctx.rubyBlockId);
+                unconditionalJump(current, rescueHeaderBlock, cctx.inWhat, a->loc);
+                cctx.rescueScope = rescueHeaderBlock;
 
                 // We have a simplified view of the control flow here but in
                 // practise it has been reasonable on our codebase.
                 // We don't model that each expression in the `body` or `else` could
                 // throw, instead we model only never running anything in the
                 // body, or running the whole thing. To do this we  have a magic
                 // Unanalyzable variable at the top of the body using
-                // `rescueStartTemp` and one at the end of the else using
+                // `exceptionValue` and one at the end of the else using
                 // `rescueEndTemp` which can jump into the rescue handlers.
-                auto rescueHandlersBlock = cctx.inWhat.freshBlock(cctx.loops, cctx.rubyBlockId);
-                auto bodyBlock = cctx.inWhat.freshBlock(cctx.loops, cctx.rubyBlockId);
-                auto rescueStartTemp = cctx.newTemporary(core::Names::rescueStartTemp());
-                synthesizeExpr(rescueStartBlock, rescueStartTemp, what->loc, make_unique<Unanalyzable>());
-                conditionalJump(rescueStartBlock, rescueStartTemp, rescueHandlersBlock, bodyBlock, cctx.inWhat, a->loc);
+                auto rescueHandlersBlock = cctx.inWhat.freshBlock(cctx.loops, handlersRubyBlockId);
+                auto bodyBlock = cctx.inWhat.freshBlock(cctx.loops, bodyRubyBlockId);
+                auto exceptionValue = cctx.newTemporary(core::Names::exceptionValue());
+                synthesizeExpr(rescueHeaderBlock, exceptionValue, what->loc, make_unique<GetCurrentException>());
+                conditionalJump(rescueHeaderBlock, exceptionValue, rescueHandlersBlock, bodyBlock, cctx.inWhat, a->loc);
 
                 // cctx.loops += 1; // should formally be here but this makes us report a lot of false errors
                 bodyBlock = walk(cctx, a->body.get(), bodyBlock);
-                auto elseBody = cctx.inWhat.freshBlock(cctx.loops, cctx.rubyBlockId);
+                // TODO: Do we need to insert a sentinel to distinguish 'else' from body?
+                auto elseBody = cctx.inWhat.freshBlock(cctx.loops, elseRubyBlockId);
                 unconditionalJump(bodyBlock, elseBody, cctx.inWhat, a->loc);
```

Comment:
Should this also be conditional on the result of GetCurrentException at the end of the body block?

Reply:
I think this question is answered by the TODO above.